### DISCUSSION TITLE
Expose mutation pending input for better optimistic updates

### DISCUSSION THREAD

**Discussion by nirtamir2** [TanStack/query#4783] [Ideas] [answered]
Hi
I would like to suggest a feature I get from SolidStart - expose mutation pending input for better optimistic updates.
See https://start.solidjs.com/api/createRouteAction is the optimistic UI section.
In SolidStart you can get the mutation input for pending mutations.
This is useful in representing more granular optimistic updates, and it shines when we have more than a single query that is in flight. The user can get the pending input of the mutations and show them in the UI, and when the mutation request finish - the matching pending input is deleted.

You can see an example here https://github.com/solidjs/solid-start/blob/main/examples/todomvc/src/routes/index.tsx#L133

We can show the in-flight new to-do items easily.
```tsx
  const [addingTodo, addTodo] = createServerMultiAction$(addTodoFn);
...
// Show the in-fight todos when adding them
 <For each={addingTodo}>
            {sub => (
              <li class="todo pending">
                <div class="view">
                  <label>{sub.input.get("title") as string}</label>
                  <button disabled class="destroy" />
                </div>
              </li>
            )}
          </For>

```

This approach moves the responsibility of optimistic-update UI to the UI itself, instead of the `queryClient` data layer.  Which leads to a state that is easier to maintain and easier to debug.

If I would do it with @tanerstack/query, I would have to update the cache optimistically myself and handle success/error actions too. And for multi-action requests - this will be harder - because some of the mutations can succeed and some of them fail.

I think that `useMutation` can return some input property that equals the mutation input of the inflight mutation.
We can also expose `inputs` for all the in-flight request inputs for the same mutation.

We may need to expose `useMultipleMutation` / add an option to `useMutation` to mark if this pending input property will take the latest

Solid `createRouteMultiAction` returns a tuple of the mutation details and the mutation mutate function.

## Suggestion API 1 (worse DX. but easier to migrate)

```ts
async function fn(param: FnParam){
 return [] as FnReturnType
}

const mutation = useMutation(fn)
mutation.input // returns FnParam 
mutation.inputs // returns Array<FnParam> 

```

## Suggestion API 2 (better DX, harder to migrate)

```ts
async function fn(param: FnParam){
 return [] as FnReturnType
}

const [mutationDetails, mutateFn] = useMutation(fn)
mutationDetails.input // returns FnParam 
mutationDetails.error
mutationDetails.isPending  //etc...
```
and
```ts
const [mutations, mutateFn] = useMultiMutation(fn)

mutations.map(mutationDetails => {
 mutationDetails.input // returns FnParam 
 mutationDetails.error
 mutationDetails.isPending  //etc...

)

```

For example.

I'm not sure about the exact API I want (I think solid.js mutation API is better but it requires major breaking change) but I want to enable the option to get the mutation input in order to handle the optimistic updates in the UI instead of the `queryClient`.

WDYT?

**Comment by TkDodo** [comment, marked as answer]
I'm not against this, and it's not too hard to do, but I don't fully understand the value yet. This assumes that the query and mutation live in the same component, which imo is rarely the case. For example, the mutation can live in a dialog while the list that renders the query lives in a separate component. 

So we would need a way to expose the input not only as a returned value from useMutation, but maybe also generally via another hook, similar to `useIsMutating`

**Reply by nirtamir2** [reply]
Hi @TkDodo 
Thanks for your quick response!
Most optimistic updates happen when the query is active, and we have a mutation somewhere that updates this query data. You are right - they do not have to be in the same component. So yes - we can expose it differently (like `useIsMutating`) or we can use a context in the userland to use the same instance of `useMutation` and expose the pending from the mutation itself. But notice that if we want to expose `usePendingData` hook - we need to link it to the matched mutation somehow - maybe via key param. When there is a dialog - I usually add a loading spinner in the submit button, because we can have error and I don't want the user will lose its current context / close the dialog. I usually patch the cache optimistically when the query and the mutation in the same view

The value is that we don't have to update the data in the data layer for optimistic updates - but in the view layer. You get the input and render it in the component, instead of updating the queryClient with `onMutate` and then again updating it `onSuccess` / `onError`. Because this optimistic update data is not the "true" data. 
Another benefit of this is by supporting multi-action mutation. If I update the `queryClient` by adding 3 todos with a fake id from the client and one of the requests fails - I would like to find this specific id and remove it from my todos, but I still want to show the other 2 fake ids that are in-flight. `usePendingData` separate hook (like is mutating) can be a good option, but we need to link it somehow to the right mutation (I don't know how we can do it). Do you mean by adding a key per-mutation and searching by this key?


Example of how we currently handle it with react query (trpc just for the types):
https://github.com/nirtamir2/query-pending-example/blob/main/src/pages/index.tsx
```tsx
import { type NextPage } from "next";
import { api } from "../utils/api";

function useAddNoteMutation() {
  const utils = api.useContext();
  return api.notes.createNote.useMutation({
    onMutate: (variables) => {
      const fakeId = `fake_${Math.random()}`;
      utils.notes.getNotes.setData(undefined, (prevNotes) => {
        if (prevNotes == null) {
          return prevNotes;
        }
        return [...prevNotes, { id: fakeId, text: variables.name }];
      });
      return { fakeId };
    },
    onSuccess: (newTodo, variables, context) => {
      utils.notes.getNotes.setData(undefined, (prevNotes) => {
        if (prevNotes == null) {
          return prevNotes;
        }
        return [
          ...prevNotes.filter((note) => note.id !== context?.fakeId),
          newTodo,
        ];
      });
    },
    onError: (error, variables, context) => {
      utils.notes.getNotes.setData(undefined, (prevNotes) => {
        if (prevNotes == null) {
          return prevNotes;
        }
        return prevNotes.filter((note) => note.id !== context?.fakeId);
      });
    },
  });
}

const Home: NextPage = () => {
  const notesQuery = api.notes.getNotes.useQuery();

  const addNoteMutation = useAddNoteMutation();

  return (
    <>
      <main className="flex min-h-screen flex-col items-center justify-center">
        <div>
          {notesQuery.data?.map((note) => (
            <div key={note.id}>{note.id} {note.text}</div>
          ))}
        </div>
        <form
          onSubmit={(event) => {
            event.preventDefault();
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-expect-error
            const name = event.currentTarget.name.value as string;

            addNoteMutation.mutate({ name });
            event.currentTarget.reset();
          }}
        >
          <input className="rounded border" name="name" type="text" />
          <button type="submit">+</button>
        </form>
      </main>
    </>
  );
};

export default Home;
```


Example of how I want it to behave (not working):
https://github.com/nirtamir2/query-pending-example/blob/main/src/pages/with-query-inputs.tsx
```tsx
import { type NextPage } from "next";

import { api } from "../utils/api";


const Home: NextPage = () => {
  const notesQuery = api.notes.getNotes.useQuery();
  const addNoteMutation = api.notes.createNote.useMutation();
  
  return (
    <>
      <main className="flex min-h-screen flex-col items-center justify-center">
        <div>
          {notesQuery.data?.map((note) => (
            <div key={note.id}>{note.id} {note.text}</div>
          ))}
          {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
          {/* @ts-ignore */}
          {(addNoteMutation.?inputs as Array<string> | undefined)?.map(fakeTodo =>
            <div key={fakeTodo}>{fakeTodo}</div>
          )}
        </div>
        <form
          onSubmit={(event) => {
            event.preventDefault();
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-expect-error
            const name = event.currentTarget.name.value as string;

            addNoteMutation.mutate({ name });
            event.currentTarget.reset();
          }}
        >
          <input className="rounded border" name="name" type="text" />
          <button type="submit">+</button>
        </form>
      </main>
    </>
  );
};

export default Home;

```

I can replace the `addNoteMutation.inputs` in `usePendingData("some-link-to-mutation")` if this makes more sense for you. But the main thing is the ability to get the pending mutation inputs.
If you want more motivation for the pending mutation input - you can watch https://youtu.be/eAwuPvRXNdY?t=5958 .
If we get the pending mutation data, Instead of patching the cache - we do the optimistic update in the component itself.

**Reply by TkDodo** [reply]
I think `usePendingData({ mutationKey: ['my', 'mutation'] })` would work if you provide that same key to your mutation. We _could_ return it also from `useMutation`, but keep in mind that if you call `.mutate` twice, you would only get the _latest_ data returned - same as you can only track the `loading` state of the latest invocation. If we go with the separate hook, the returned `inputs` would likely be untyped

**Reply by nirtamir2** [reply]
@TkDodo Yes - this would be amazing.
If I understand you correctly - the `useMutation` will return input of only the last invocation (or undefined if the no pending mutations), and if I would like to get all in-flight invocations - I would use `usePendingData` (or similar new hook) that will return an array of the pending inputs. 
This API looks good to me. If we compare it to solid-start - they created `createdRouteAction` (for single pending input) or `createRouteMultiAction` for multiple actions. https://start.solidjs.com/api/createRouteMultiAction but I think it's ok to keep it as you suggested.
It's ok for me to use a hook like `usePendingData` untyped as long as we can create a layer over it that will make it type-safe (meaning type-safe in `trpc`). Like the query client in `trpc` that is type-safe, even when `tanerstack/query` is not type-safe.
WDYT?

**Reply by TkDodo** [reply]
yeah I'll definitely think about it for v5. What should definitely be possible is to add the `input` to be returned form `useMutation`. The separate hook could theoretically be built in user-land - maybe we'll just provide instructions for that

**Reply by TkDodo** [reply]
actually - we already expose `variables` 🎉 

https://www.typescriptlang.org/play?#code/ATCWFsAcHsCcBdgG9gFcDOBTAsq+BDeUaAO2AF9gAzWacYAcgAECT0CBjAawHpZN8HeAFoAjqkywAngwBQIecA6l2yYADd8sUPgBGAG0zoKwALxosuAkVIAKJIpDg8hYiQBiJAFzBboEpB4PuzaJADmAJRmAHzAAAq04KBYAHT86ND66pi2AKwRiuQFispsmZgp+tBhtpraeoboBSA8PCDtHcAAegD8QA

**Reply by TkDodo** [reply]
and if you'd need a a separate hook, building something like [useIsMutating](https://github.com/TanStack/query/blob/eeec5f77bc9a703ffb6a6d283dcedada34aa3c75/packages/react-query/src/useIsMutating.ts#L11-L27) should be possible on your end right now as well.

**Reply by nirtamir2** [reply]
Amazing! I will try it!

**Reply by TkDodo** [reply]
I noticed that variables get exposed but they are not reset when the mutation succeeds. We might ship this for v5 though, stay tuned. Thank you for putting the idea in my head 🙌

**Reply by TkDodo** [reply]
shipped it for v5 🎉

**Reply by nirtamir2** [reply]
Awsome!
